### PR TITLE
Additional Functionality & General Updates (Thermostat, Query Behavior)

### PR DESCRIPTION
**Pull request by skimdankish**
@mdopp If you agree now you can also withdraw your PR as we had discussed.

@marziman :

I have now squashed the commits from mine and @mdopp - unfortunately, I was unable to merge #22 as I don't fully understand the changes.

The tag issue seemed to be a simple fix and I saw some people were having this issue because of misspelling and such.  Let me know if you have any questions, here is the synopsis below:

I am currently running on this code with no errors and have tried upwards, downwards, backwards, and forwards to try to "trick" my commands using Google Cloud Console and REST commands.  So far so good.

Highlights:
- Adds Thermostat functionality (tested with heating, and manually adjusted to emulate)
- Ignores misspelled or unsupported tags (should not crash on sync)
- You can dim generically without specifying by how much: e.g. "Dim [x] lights" instead of "Dim [x] lights 50%".  Google controls the interval on all generic commands.

Changelog:
- UPDATE -> Split Query and Execute to separate functions
- UPDATE -> Query Function retrieves item data, then process response by trait (per @mdopp)
- UPDATE -> Execute function goes also by trait (per @mdopp)
- NEW -> getLightData: retrieves lighting attributes OH item (on/off for switch lights and also brightness 0-100 for dimmers)
- NEW -> getSwitchData: retrieves switch attriutes from OH item (on/off)
- NEW -> getColorData: retrieves color attributes from OH item, SpectrumRGB
- NEW -> getTempData: Google requires to first query thermostat before executing.  Therefore this code will provide appropriate response based on query.
- UPDATE -> getThermostatItems: fixed to allow for CurrentTemperature tag (standalone sensor/item without Thermostat group)
- UPDATE -> adjustTemperatureWithItems: Fixed isF (function from utils is from Alexa code)
- UPDATE -> syncAndDiscoverDevices: Named loop to allow switch default to break out to avoid crashing on sync on mismatch item tags
- UPDATE -> syncAndDiscoverDevices: Google wants deviceType as string, removed brackets as it was a soft-warning from their reference docs
- UPDATE -> getSwitchableTraits: Allow for Temperature and Thermostat to handle Query-Trait functionality (per @mdopp)

### CODE REVIEW

**Comment by skimdankish** [comment]
This would resolve #19 , #21, and #23

File: functions/openhab.js [hunk 1]
```diff
@@ -141,16 +143,113 @@ function getItemsState(request,response) {
...
+
+	return thermData;
+}
+
+/**
+ *  Retrieves Light Attributes from OpenHAB Item
+ **/
+function getLightData(item) {
+	return {
+		on: item.state === 'ON' ? true : (item.state === 0 ? false : true),
```
Comment (sk8ter) [inline functions/openhab.js]: This fails for me, since state could be string. See my comment: https://github.com/openhab/openhab-google-assistant/pull/24#issuecomment-368660201

Comment (skimdankish) [reply functions/openhab.js]: Thanks I see what you mean, if we change to Number(item.state) === 0 I have confirmed it fixes the issue.

Thanks for checking!  Do you also try the thermostat functions?  I did my best to test them manually since I only have thermostat via heat.

Comment (sk8ter) [reply functions/openhab.js]: No, unfortunately I don't have thermostats yet.

**Comment by marziman** [comment]
@skimdankish ,

many thanks to you and @mdopp for your efforts.
I wanted to merge it now, but there are now some conflicts. Can you please take a look?
I think this is related to the merged code cleanup PR (without big logical changes).
Thank you and great job!

BR Mehmet

**Comment by skimdankish** [comment]
Sure, I will adjust and run it on my setup Monday/Tuesday.

Ryan

**Comment by skimdankish** [comment]
@marziman 

Most of the changes were ok to keep, just the split Query/Execute (1st Conflict) that had to be overridden.

I ran threw a bunch of testing and function is still working well and efficiently. Let me know if you have further questions or issues!

Ryan

**Comment by eimix** [comment]
Hello,

looking forward for this improvements.
But could not find/understand how to add this new functionality (after pull) to my setup:

Openhab 2.2.0 +
https://myopenhab.org/ +
Google assistant (it has connection to myopenhab.org)

Does this update should be applied to https://myopenhab.org/? (does this _pull_ upgrades myopenhab.org?)
should i setup my one **openhab-cloud** to get newest functions from **openhab-google-assistant**?

**Comment by skimdankish** [comment]
@eimix 

TL;DR >> The function found here can ONLY work with your own openhab-cloud instance.

It seems like you are using the "production" Openhab Action/Function which is LIVE.  The part I am working on is only for development, so in order for these updates to be incorporated:
- The Pull Request needs to be approved and merged into the official development of openhab-google-assistant
- The development would need to be merged/replace the existing production Openhab Action/Function (this may require some approval and/or testing from Google, I don't know...)

@marziman  May be able to answer some further questions as I am not familiar with the process...

Thanks,
Ryan

**Comment by skimdankish** [comment]
@marziman 

Just checking in on this...I had done my review and made sure there are no conflicts.  I am now running on what the merged code with no error.  

The one thing that failed is the tagging fix - the loop escape I coded doesn't quite do the trick - perhaps someone more skilled than I can fix that...until then people will need to know which tags will work

This PR will allow for Temperature/Humidity/Thermostat tags, so that should fix the Alexa/Google Home contingent of OpenHAB Users.

Thanks and let me know what else I can do here.

Ryan

**Comment by marziman** [comment]
Will test it the upcoming days. Sorry for the delay!
BR Mehmet

**Comment by tagazok** [comment]
Hi,
Is there any news on this PR? :)

**Comment by ahielg** [comment]
Hello,
Can someone please approve this PR?

**Comment by mvalla** [comment]
Yes this PR adds a lot and solves few issues. Can anyone else approve it and push it on the myopenhab cloud? I think GH integration is becoming very popular for OpenHAB and deserves more people authorized to approve/deploy

**Comment by marziman** [comment]
Hi,
I made a reviewof this PR.
It is not just like a review and deploying.
I also need to trigger the functionupdate at Google. So I am going to retest it and deploy it.

BR Mehmet

**Comment by marziman** [comment]
@skimdankish,

going to merge. My only issue is that we have not written testcases for this, but this needs to be improved at other places too. I will ping you soon, cause I will setup better testing and s CI for easy deployments.

Thx & BR Mehmet

**Comment by skimdankish** [comment]
@marziman 

Sure, let me know what is needed for test cases and I'll help where I can!

Ryan

**Comment by MichaelFink** [comment]
Was this change already pushed to the prod system? I'm still experiencing issues which should have been solved by this change. Like unknown state for dimmers and issues with current temperature tags.

**Comment by skimdankish** [comment]
@MichaelFink 

Which do you mean by "prod system"?  This is now in the dev version in Github, but I don't think it was pushed to Google...

**Comment by MichaelFink** [comment]
Yep, with prod system I mean the "Actions on Google" project where the code is running. 
It's a bit sad that there is quite some delay until new versions get pushed. Happy to help out in case this helps :-)

- Michael

**Comment by marziman** [comment]
Hi,
I will (as always) ping Google.
Will get back to you soon @MichaelFink

Sure, your help is highly welcome. Lets take some time soon and we can check where you might go in?

BR Mehmet

**Comment by MichaelFink** [comment]
@marziman Perhaps I can help with that. I'm working for this company :-) Do you have a ticket id or is this happening via email?

**Comment by marziman** [comment]
Hey guys,

I ve the same problem as last time. I just recognized that I ve pushed the latest version to actions console. But I cant give you easy access to help testing latest deployment without sharing the overall project. That once went totally wrong, since people can see secrets etc. and we dont want to make that accessible to anyone. I can blindly press publish, since I tested it and it all worked.
There is still some small risk, that we might break something. Last time Google "broke" something...maybe you can understand my situation.

So I would go an re-publish it and ping Google by mail. @MichaelFink yes correct, it is by mail.
Since openHAB is completely different and has "a special process" :-)

**Comment by marziman** [comment]
Hi,

maybe I found a way. @skimdankish can you maybe test it also with the official openHAB integration which uses myopenhab.org instead of your own deployment? I dont have a Thermostat setup atm to test fully. In my tests it worked fine so far. 

Thanks!

**Comment by MichaelFink** [comment]
Yep, sharing access to the project widely is probably not a good idea.

What if we would ask Google if we could get another listing for openhab with a separate project? "openhab Beta" or something. Then you could always deploy the latest version to that new project and people which want to life on the edge can use that one. If no issue pop up, you deploy the same version to the regular project a few days or weeks later.
With that it's easier for people to test new versions and you get more traffic and can be more confident that it doesn't break anything.

**Comment by marziman** [comment]
Great!
Did you hit the tagging issue, too? If yes, this should be also solved.

**Comment by MichaelFink** [comment]
Yes, for the temperature sensors I'm using the "CurrentTemperature" tag. This didn't work before (syncing devices wasn't possible). Now it works!
Also dimming now works because Google Assistant can query the current status of the lamp. Nice job!

**Comment by skimdankish** [comment]
What is all this productivity on a Friday Eve/Sat AM?!?

My biggest question....does Celsius work...?

Spent quite a bit of time on CurrentTemp standalone bc I have a temperature sensors which are not necessarily thermostats.

**Comment by MichaelFink** [comment]
@skimdankish Same here. Just CurrentTemperature tags. Celsius works fine.

**Comment by marziman** [comment]
@skimdankish actually big kudos to you and thanks for your contributions.
I am more active in the night, when the baby sleeps ;-) and I am not busy with other things.

**Comment by skimdankish** [comment]
@marziman

Funny you say that as much of my motivation and inspiration was hands free scenes since my little guy is quite obsessed with "uppy".

You still need me to test general production? I like my own deployment bc such low latency ;)

@MichaelFink I'm very glad the American didn't break a unit functionality 95% of the world uses...
